### ISSUE TITLE
Job Handler Reboot Failure or Missing Documentation

### ISSUE THREAD

**Issue by philcali** [awslabs/aws-iot-device-client#338]
**Describe the bug**

I'm confused about the `reboot.sh` script and the jobs feature. On my devices, I have observed that the first reboot job received by the device will hit this block: https://github.com/awslabs/aws-iot-device-client/blob/main/sample-job-handlers/reboot.sh#L35 and then will promptly remove the lock file. So I attempted to remove the line and add a retry config to have the job execution cycle twice on the reboot:

1. to trigger the reboot and write the lock
2. to run again and confirm that the device reboot with lock file present

It is confusing what is expected behavior. Either there is an error in my job configuration or an error in the script, but I don't know which is which.

**To Reproduce**

1. install the `aws-iot-device-client`
2. start job feature and reboot scripts
3. create a job with the reboot document
4. Observe the `FAILED` execution for the device.

**Expected behavior**

I'm not sure: I guess it makes sense the way the script is written to expect a retry.

**Actual behavior**

Unable to get the current `reboot.sh` script to function without modification

**Logs**

Removed: https://github.com/awslabs/aws-iot-device-client/blob/main/sample-job-handlers/reboot.sh#L35

```
$ journalctl -r -u aws-iot-device-client
Oct 05 20:00:15 raspberrypi aws-iot-device-client[501]: 2022-10-06T00:00:15.441Z [INFO]  {JobsFeature.cpp}: No pending jobs are scheduled, waiting for the next incoming job
Oct 05 20:00:14 raspberrypi aws-iot-device-client[501]: 2022-10-06T00:00:14.484Z [INFO]  {JobsFeature.cpp}: Job executed successfully!
Oct 05 20:00:14 raspberrypi aws-iot-device-client[501]: 2022-10-06T00:00:14.484Z [INFO]  {JobsFeature.cpp}: Job exited with status: 0
Oct 05 20:00:14 raspberrypi sudo[529]: pam_unix(sudo:session): session closed for user root
Oct 05 20:00:14 raspberrypi sudo[529]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
Oct 05 20:00:14 raspberrypi sudo[529]:     root : PWD=/ ; USER=root ; COMMAND=/usr/bin/test -f /var/tmp/dc-reboot.lock
Oct 05 20:00:13 raspberrypi aws-iot-device-client[501]: 2022-10-06T00:00:13.572Z [INFO]  {JobEngine.cpp}: About to execute: /usr/share/pinthesky/job/handlers/reboot.sh root
Oct 05 20:00:13 raspberrypi aws-iot-device-client[501]: 2022-10-06T00:00:13.570Z [INFO]  {JobEngine.cpp}: Did not find any arguments in the incoming job document. Value should be a JSON array of arguments
Oct 05 20:00:13 raspberrypi aws-iot-device-client[501]: 2022-10-06T00:00:13.560Z [INFO]  {JobEngine.cpp}: About to execute step with name: Reboot Camera
-- Boot 8e47899903e04bacbbc3a9376201ac18 --
Oct 05 19:58:57 raspberrypi systemd[1]: aws-iot-device-client.service: Consumed 1min 10.426s CPU time.
Oct 05 19:58:57 raspberrypi systemd[1]: Stopped AWS IoT Device Client.
Oct 05 19:58:57 raspberrypi systemd[1]: aws-iot-device-client.service: Succeeded.
Oct 05 19:58:56 raspberrypi systemd[1]: Stopping AWS IoT Device Client...
Oct 05 19:58:56 raspberrypi aws-iot-device-client[501]: 2022-10-05T23:58:56.492Z [WARN]  {JobsFeature.cpp}: Job execution failed!
Oct 05 19:58:56 raspberrypi aws-iot-device-client[501]: 2022-10-05T23:58:56.491Z [INFO]  {JobsFeature.cpp}: Job exited with status: 1
Oct 05 19:58:56 raspberrypi sudo[5632]: pam_unix(sudo:session): session closed for user root
Oct 05 19:58:55 raspberrypi sudo[5632]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
Oct 05 19:58:55 raspberrypi sudo[5632]:     root : PWD=/ ; USER=root ; COMMAND=/usr/sbin/reboot
Oct 05 19:58:55 raspberrypi sudo[5629]: pam_unix(sudo:session): session closed for user root
Oct 05 19:58:55 raspberrypi sudo[5629]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=0)
Oct 05 19:58:55 raspberrypi sudo[5629]:     root : PWD=/ ; USER=root ; COMMAND=/usr/bin/test -f /var/tmp/dc-reboot.lock
Oct 05 19:58:55 raspberrypi aws-iot-device-client[501]: 2022-10-05T23:58:55.475Z [INFO]  {JobEngine.cpp}: About to execute: /usr/share/pinthesky/job/handlers/reboot.sh root
Oct 05 19:58:55 raspberrypi aws-iot-device-client[501]: 2022-10-05T23:58:55.475Z [INFO]  {JobEngine.cpp}: Did not find any arguments in the incoming job document. Value should be a JSON array of arguments
Oct 05 19:58:55 raspberrypi aws-iot-device-client[501]: 2022-10-05T23:58:55.475Z [INFO]  {JobEngine.cpp}: About to execute step with name: Reboot Camera
```

```
$ aws iot list-job-executions-for-job --job-id 57d6ab2e-8bfb-48d0-9285-5834a485da85
{
    "executionSummaries": [
        {
            "thingArn": "arn:aws:iot:REGION:ACCOUNT:thing/PitsCamera3",
            "jobExecutionSummary": {
                "status": "SUCCEEDED",
                "queuedAt": "2022-10-05T19:58:57.375000-04:00",
                "startedAt": "2022-10-05T20:00:13.514000-04:00",
                "lastUpdatedAt": "2022-10-05T20:00:14.511000-04:00",
                "executionNumber": 2,
                "retryAttempt": 1
            }
        },
        {
            "thingArn": "arn:aws:iot:REGION:ACCOUNT:thing/PitsCamera3",
            "jobExecutionSummary": {
                "status": "FAILED",
                "queuedAt": "2022-10-05T19:58:54.967000-04:00",
                "startedAt": "2022-10-05T19:58:55.542000-04:00",
                "lastUpdatedAt": "2022-10-05T19:58:56.529000-04:00",
                "executionNumber": 1,
                "retryAttempt": 0
            }
        }
    ]
}
```

If applicable, add full logs of errors and outputs to help explain your problem. Preferably, you can also [increase the verbosity of the logs](https://github.com/awslabs/aws-iot-device-client#logging).

**Environment (please complete the following information):**
 - OS: raspbian
 - Version: kernel 5.15.61+
 - Architecture: armv6
 - Device Client commit: 2790c41a954f38119cd3c87015ecf77aec28dbc8

**Additional context**

**Comment by shangabl** [comment]
Hi Philcali,

I tried this with the latest commit and with the commit you have shared and the reboot Job worked on an ubuntu machine at least.

I believe the expected behavior is that the device reboots and does not reach line 35 of the reboot.sh script. Then on restart the lock file is present. It would seem the reboot command is failing on your machine. Do you have a reboot binary in path? If you just enter reboot into the shell does it reboot?

I will sync up with my team to talk more about this. However, it might be difficult to recreate your exact environment.

**Comment by philcali** [comment]
I use the `reboot` that comes with a headless `raspbian` install for RPi (32-bit install, since this is a pi zero armv6). On these systems, the default `reboot` does not block execution, it's not immediate (thus the reason job fails with exit 1). For now I've worked worked around this issue by dropping line 35 of `reboot.sh` and adding a retry in the job configuration (though it sounds like I could drop L35 and the following the `exit 1` and achieve the same result).

I realize that recreating these issues will be difficult. I can happily trial and error things on these systems :)

**Comment by antoine-sachet** [comment]
I've got the same issue with a headless raspberry pi running debian buster 32-bit. I use the AWS-Reboot job template and all the jobs always "fail" in the console despite rebooting the device successfully.

TLDR: there's an issue with the function getting the last boot time. My workaround is to always add 5 seconds to the detected last boot time - then the reboot job succeeds.

The logs are:

```
2022-12-02T10:59:11.536Z [DEBUG] {JobsFeature.cpp}: Job ids differ
2022-12-02T10:59:11.536Z [DEBUG] {JobsFeature.cpp}: Attempting to update job execution status!
2022-12-02T10:59:11.536Z [INFO]  {JobsFeature.cpp}: Executing job: SimpleRebootNoRetry
2022-12-02T10:59:11.537Z [DEBUG] {Retry.cpp}: Retryable function starting, it will retry until success
2022-12-02T10:59:11.537Z [INFO]  {JobEngine.cpp}: About to execute step with name: Reboot
2022-12-02T10:59:11.537Z [DEBUG] {JobEngine.cpp}: Using DC default command path {/opt/hestiia/iot/bin/jobs/} for command execution
2022-12-02T10:59:11.537Z [DEBUG] {JobsFeature.cpp}: Created EphemeralPromise for ClientToken dTqVc6xmSW in the updateJobExecution promises map
2022-12-02T10:59:11.537Z [INFO]  {JobEngine.cpp}: Did not find any arguments in the incoming job document. Value should be a JSON array of arguments
2022-12-02T10:59:11.537Z [INFO]  {JobEngine.cpp}: About to execute: /opt/hestiia/iot/bin/jobs/reboot.sh  
2022-12-02T10:59:11.538Z [DEBUG] {JobEngine.cpp}: Child process now running
2022-12-02T10:59:11.538Z [DEBUG] {JobEngine.cpp}: Child process about to call execvp
2022-12-02T10:59:11.538Z [DEBUG] {JobEngine.cpp}: Parent process now running, child PID is 826
2022-12-02T10:59:11.546Z [DEBUG] {826}: Running reboot.sh
2022-12-02T10:59:11.546Z [DEBUG] {826}: Username: 
2022-12-02T10:59:11.554Z [DEBUG] {826}: username or sudo command not found
2022-12-02T10:59:11.554Z [DEBUG] {826}: Did not find /var/tmp/dc-reboot.lock
2022-12-02T10:59:11.579Z [DEBUG] {JobsFeature.cpp}: Ack received for PublishUpdateJobExecutionStatus with code {0}
2022-12-02T10:59:11.582Z [DEBUG] {JobsFeature.cpp}: Removing ClientToken dTqVc6xmSW from the updateJobExecution promises map
2022-12-02T10:59:11.583Z [DEBUG] {JobsFeature.cpp}: Success response after UpdateJobExecution for job SimpleRebootNoRetry
```

Then the device reboots in about 20 seconds - I skip to the interesting part of the logs.

```
2022-12-02T10:59:29.075Z [INFO]  {JobsFeature.cpp}: Running Jobs!
2022-12-02T10:59:29.075Z [DEBUG] {JobsFeature.cpp}: Attempting to subscribe to startNextPendingJobExecution accepted and rejected
2022-12-02T10:59:29.076Z [DEBUG] {JobsFeature.cpp}: Attempting to subscribe to nextJobChanged events
2022-12-02T10:59:29.076Z [DEBUG] {JobsFeature.cpp}: Attempting to subscribe to updateJobExecutionStatusAccepted for jobId +
2022-12-02T10:59:29.081Z [INFO]  {Main.cpp}: Client base has been notified that Secure Tunneling has started
2022-12-02T10:59:29.134Z [DEBUG] {JobsFeature.cpp}: Ack received for SubscribeToUpdateJobExecutionAccepted with code {0}
2022-12-02T10:59:29.134Z [DEBUG] {JobsFeature.cpp}: Attempting to subscribe to updateJobExecutionStatusRejected for jobId +
2022-12-02T10:59:29.153Z [DEBUG] {DeviceDefender.cpp}: SubAck: PacketId:(Device Defender), ErrorCode:0
2022-12-02T10:59:29.153Z [DEBUG] {JobsFeature.cpp}: Ack received for SubscribeToNextJobChanged with code {0}
2022-12-02T10:59:29.153Z [DEBUG] {JobsFeature.cpp}: Ack received for SubscribeToStartNextJobRejected with code {0}
2022-12-02T10:59:29.154Z [DEBUG] {JobsFeature.cpp}: Ack received for SubscribeToStartNextJobAccepted with code {0}
2022-12-02T10:59:29.181Z [DEBUG] {JobsFeature.cpp}: Ack received for SubscribeToUpdateJobExecutionRejected with code {0}
2022-12-02T10:59:29.181Z [DEBUG] {JobsFeature.cpp}: Publishing startNextPendingJobExecutionRequest
2022-12-02T10:59:29.212Z [DEBUG] {JobsFeature.cpp}: Ack received for StartNextPendingJobPub with code {0}
2022-12-02T10:59:29.216Z [DEBUG] {JobsFeature.cpp}: We have not seen a job yet, this is not a duplicate job notification
2022-12-02T10:59:29.216Z [DEBUG] {JobsFeature.cpp}: Attempting to update job execution status!
2022-12-02T10:59:29.216Z [INFO]  {JobsFeature.cpp}: Executing job: SimpleRebootNoRetry
2022-12-02T10:59:29.216Z [DEBUG] {Retry.cpp}: Retryable function starting, it will retry until success
2022-12-02T10:59:29.216Z [DEBUG] {JobsFeature.cpp}: Created EphemeralPromise for ClientToken sLG8QHdfW5 in the updateJobExecution promises map
2022-12-02T10:59:29.216Z [INFO]  {JobEngine.cpp}: About to execute step with name: Reboot
2022-12-02T10:59:29.216Z [DEBUG] {JobEngine.cpp}: Using DC default command path {/opt/hestiia/iot/bin/jobs/} for command execution
2022-12-02T10:59:29.218Z [INFO]  {JobEngine.cpp}: Did not find any arguments in the incoming job document. Value should be a JSON array of arguments
2022-12-02T10:59:29.218Z [INFO]  {JobEngine.cpp}: About to execute: /opt/hestiia/iot/bin/jobs/reboot.sh  
2022-12-02T10:59:29.218Z [DEBUG] {JobEngine.cpp}: Child process now running
2022-12-02T10:59:29.218Z [DEBUG] {JobEngine.cpp}: Child process about to call execvp
2022-12-02T10:59:29.225Z [DEBUG] {JobEngine.cpp}: Parent process now running, child PID is 582
2022-12-02T10:59:29.234Z [DEBUG] {582}: Running reboot.sh
2022-12-02T10:59:29.234Z [DEBUG] {582}: Username: 
2022-12-02T10:59:29.240Z [DEBUG] {582}: username or sudo command not found
2022-12-02T10:59:29.240Z [DEBUG] {582}: Found lock file
2022-12-02T10:59:29.249Z [DEBUG] {JobsFeature.cpp}: Ack received for PublishUpdateJobExecutionStatus with code {0}
2022-12-02T10:59:29.251Z [DEBUG] {JobsFeature.cpp}: Removing ClientToken sLG8QHdfW5 from the updateJobExecution promises map
2022-12-02T10:59:29.252Z [DEBUG] {JobsFeature.cpp}: Success response after UpdateJobExecution for job SimpleRebootNoRetry
2022-12-02T10:59:29.268Z [DEBUG] {582}: Reboot failed
2022-12-02T10:59:29.269Z [DEBUG] {JobEngine.cpp}: JobEngine finished waiting for child process, returning 256
2022-12-02T10:59:29.269Z [INFO]  {JobsFeature.cpp}: Job exited with status: 1
2022-12-02T10:59:29.269Z [WARN]  {JobsFeature.cpp}: Job execution failed!
2022-12-02T10:59:29.269Z [DEBUG] {JobsFeature.cpp}: Attempting to update job execution status!
2022-12-02T10:59:29.269Z [DEBUG] {Retry.cpp}: Retryable function starting, it will retry until success
2022-12-02T10:59:29.270Z [DEBUG] {JobsFeature.cpp}: Created EphemeralPromise for ClientToken 7n8JeT2SD6 in the updateJobExecution promises map
2022-12-02T10:59:29.315Z [DEBUG] {JobsFeature.cpp}: Ack received for PublishUpdateJobExecutionStatus with code {0}
2022-12-02T10:59:29.317Z [DEBUG] {JobsFeature.cpp}: Removing ClientToken 7n8JeT2SD6 from the updateJobExecution promises map
2022-12-02T10:59:29.318Z [DEBUG] {JobsFeature.cpp}: Success response after UpdateJobExecution for job SimpleRebootNoRetry
2022-12-02T10:59:31.138Z [INFO]  {JobsFeature.cpp}: No pending jobs are scheduled, waiting for the next incoming job
```

So the execution logs of `reboot.sh` are:
```
# BEFORE REBOOT
2022-12-02T10:59:11.546Z [DEBUG] {826}: Running reboot.sh
2022-12-02T10:59:11.546Z [DEBUG] {826}: Username: 
2022-12-02T10:59:11.554Z [DEBUG] {826}: username or sudo command not found
2022-12-02T10:59:11.554Z [DEBUG] {826}: Did not find /var/tmp/dc-reboot.lock
```

```
# AFTER REBOOT
2022-12-02T10:59:29.234Z [DEBUG] {582}: Running reboot.sh
2022-12-02T10:59:29.234Z [DEBUG] {582}: Username: 
2022-12-02T10:59:29.240Z [DEBUG] {582}: username or sudo command not found
2022-12-02T10:59:29.240Z [DEBUG] {582}: Found lock file
2022-12-02T10:59:29.268Z [DEBUG] {582}: Reboot failed
```

Clearly there's something wrong as the reboot was successful.

To understand why the reboot failed according the lock file, I added more logging to reboot.sh:

```
if test -f "$REBOOT_LOCK_FILE"; then
    echo "Found lock file"
    LOCK_FILE_BOOT=$(cat $REBOOT_LOCK_FILE)
    rm $REBOOT_LOCK_FILE
    get_last_boot_time
    echo LOCK_FILE_BOOT=$LOCK_FILE_BOOT
    echo GLOBAL_LAST_BOOT=$GLOBAL_LAST_BOOT
    if [ "$LOCK_FILE_BOOT" -lt "$GLOBAL_LAST_BOOT" ]; then
      echo "Reboot was successful"
      exit 0
    else
      echo "Reboot failed"
      exit 1
    fi
 ```
 
 and I ran the job again.
 
 After the second reboot, the execution logs were:
 ```
 2022-12-02T11:15:28.023Z [DEBUG] {574}: Running reboot.sh
2022-12-02T11:15:28.023Z [DEBUG] {574}: Username: 
2022-12-02T11:15:28.030Z [DEBUG] {574}: username or sudo command not found
2022-12-02T11:15:28.030Z [DEBUG] {574}: Found lock file
2022-12-02T11:15:28.060Z [DEBUG] {574}: LOCK_FILE_BOOT=1669979710
2022-12-02T11:15:28.060Z [DEBUG] {574}: GLOBAL_LAST_BOOT=1669979709
2022-12-02T11:15:28.060Z [DEBUG] {574}: Reboot failed
```

Obviously the `LOCK_FILE_BOOT` is not less than `GLOBAL_LAST_BOOT` so it fails - but why? I did just reboot so `GLOBAL_LAST_BOOT` should be about 20 seconds higher than `LOCK_FILE_BOOT`.

This is the function used by `reboot.sh` to get the last boot time (with an added echo)

```
get_last_boot_time() {
  # Calculate exact time of last boot
  UPTIME=$(cat /proc/uptime | head -n1 | awk '{print $1;}')
  CURRENT=$(date "+%s")
  GLOBAL_LAST_BOOT=$(awk -v currentTime=$CURRENT -v uptime=$UPTIME 'BEGIN {printf "%.0f",currentTime-uptime}')
  echo $GLOBAL_LAST_BOOT
}

```


```
$ get_last_boot_time
1669979721
```

This value is correct (about 20s after the reboot) and would have triggered a successful response. But the reboot script got  `GLOBAL_LAST_BOOT=1669979709` when it ran after the reboot.

So it looks to me like the issue is that `get_last_boot_time` does NOT work properly immediately after booting.  I don't know how to fix the function `get_last_boot_time`, but my workaround is to give 5 seconds of leeway with `GLOBAL_LAST_BOOT=$(($GLOBAL_LAST_BOOT + 5))`.

```
if test -f "$REBOOT_LOCK_FILE"; then
    echo "Found lock file"
    LOCK_FILE_BOOT=$(cat $REBOOT_LOCK_FILE)
    rm $REBOOT_LOCK_FILE
    get_last_boot_time
    echo LOCK_FILE_BOOT=$LOCK_FILE_BOOT
    echo GLOBAL_LAST_BOOT=$GLOBAL_LAST_BOOT
    # Workaround (cf issue #338)
    GLOBAL_LAST_BOOT=$(($GLOBAL_LAST_BOOT + 5))
    echo GLOBAL_LAST_BOOT_FIXED=$GLOBAL_LAST_BOOT
    if [ "$LOCK_FILE_BOOT" -lt "$GLOBAL_LAST_BOOT" ]; then
      echo "Reboot was successful"
      exit 0
    else
      echo "Reboot failed"
      exit 1
    fi
```

It's stupid but it works and my remote job is finally successful!

```
2022-12-02T11:34:01.492Z [DEBUG] {580}: Running reboot.sh
2022-12-02T11:34:01.492Z [DEBUG] {580}: Username: 
2022-12-02T11:34:01.498Z [DEBUG] {580}: username or sudo command not found
2022-12-02T11:34:01.498Z [DEBUG] {580}: Found lock file
2022-12-02T11:34:01.521Z [DEBUG] {580}: LOCK_FILE_BOOT=1669980822
2022-12-02T11:34:01.521Z [DEBUG] {580}: GLOBAL_LAST_BOOT=1669980820
2022-12-02T11:34:01.521Z [DEBUG] {580}: GLOBAL_LAST_BOOT_FIXED=1669980825
2022-12-02T11:34:01.521Z [DEBUG] {580}: Reboot was successful
```

**Comment by shangabl** [comment]
Hi Philcali,

Sorry for the delayed response here I forgot to reply after thinking about this issue originally. So, yes currently this script isn't going to work if reboot doesn't block execution... I'm glad you were able to make some modifications to the job handlers to fit your system though that is a big reason why this is an open source project. Can you share the code that works for your system now? I can then test that on other platforms and if it works all around perhaps we can make those changes to the reboot script.

**Comment by shangabl** [comment]
Hi antoine-sachet,

Thank you so much for the bringing this to our attention. I really appreciate you diving into the issue and providing that level of detail. That is very interesting I will speak to my team about why the time is incorrect here and attempt to fix.

Thanks again for reporting this!

**Comment by shangabl** [comment]
Fixed the reboot.sh Job handler so it won't fail the job if the execution isn't blocked.
